### ISSUE TITLE
1.2.0: LANG=de_DE@euro crashes borg, works fine with 1.1.17.

### ISSUE THREAD

**Issue by slashbeast** [borgbackup/borg#6406]
## Have you checked borgbackup docs, FAQ, and open Github issues?

Yes

## Is this a BUG / ISSUE report or a QUESTION?

BUG

#### Your borg version (borg -V).

borg 1.2.0

#### Operating system (distribution) and version.

Gentoo ~amd64

#### How much data is handled by borg?

-

#### Full borg commandline that lead to the problem (leave away excludes and passwords)

borg

## Describe the problem you're observing.

The borg 1.2.0, unlike 1.1.17, crashes on locale set to de_DE@euro (ISO locale).
```
Traceback (most recent call last):
  File "/usr/lib/python-exec/python3.9/borg", line 33, in <module>
    sys.exit(load_entry_point('borgbackup==1.2.0', 'console_scripts', 'borg')())
  File "/usr/lib/python3.9/site-packages/borg/archiver.py", line 5077, in main
    args = archiver.get_args(sys.argv, os.environ.get('SSH_ORIGINAL_COMMAND'))
  File "/usr/lib/python3.9/site-packages/borg/archiver.py", line 4869, in get_args
    result = self.parse_args(argv[1:])
  File "/usr/lib/python3.9/site-packages/borg/archiver.py", line 4914, in parse_args
    parser = self.build_parser()
  File "/usr/lib/python3.9/site-packages/borg/archiver.py", line 4339, in build_parser
    """) + ItemFormatter.keys_help()
  File "/usr/lib/python3.9/site-packages/borg/helpers/parseformat.py", line 733, in keys_help
    keys = cls.available_keys()
  File "/usr/lib/python3.9/site-packages/borg/helpers/parseformat.py", line 727, in available_keys
    keys.extend(formatter.get_item_data(fake_item).keys())
  File "/usr/lib/python3.9/site-packages/borg/helpers/parseformat.py", line 797, in get_item_data
    source = item.get('source', '')
SystemError: <cyfunction PropDict._make_property.<locals>._get at 0x7f5b4a2dee10> returned NULL without setting an error
```
reported on https://bugs.gentoo.org/834575

#### Can you reproduce the problem? If so, describe how. If not, describe troubleshooting steps you took before opening the issue.

Just run 'borg' with locale set to 'de_DE@euro'

**Comment by ThomasWaldmann** [comment]
@slashbeast can you tell which Cython version was used?

**Comment by pete-dawgg** [comment]
i experienced the same bug on similar systems (gentoo ~amd64) and in my case it's not LANG but LC_ALL that causes this error (i created above bugreport for gentoo).
running borg with any (?) LANG with LC_ALL unset  works. i tested "de_DE@euro" "de_DE.iso885915@euro" "de_DE.iso88591" and "de_DE.utf8"
Cython version is cython-0.29.28 (from portage dev-python/cython-0.29.28

**Comment by ThomasWaldmann** [comment]
TODO: reproduce this on debian or ubuntu.

**Comment by pete-dawgg** [comment]
Error reproducible with Devuan Chimaera  (like Debian Bullseye). The precompiled borg-linux64-executable works with 
LANG="de_DE.iso885915@euro" but fails with LC_ALL="de_DE.iso885915@euro":
```
LC_ALL="de_DE.iso885915@euro" ./borg-linux64 
Traceback (most recent call last):
  File "borg/__main__.py", line 15, in <module>
  File "borg/archiver.py", line 5077, in main
  File "borg/archiver.py", line 4869, in get_args
  File "borg/archiver.py", line 4914, in parse_args
  File "borg/archiver.py", line 4339, in build_parser
  File "borg/helpers/parseformat.py", line 733, in keys_help
  File "borg/helpers/parseformat.py", line 727, in available_keys
  File "borg/helpers/parseformat.py", line 797, in get_item_data
SystemError: <cyfunction PropDict._make_property.<locals>._get at 0x7f4f57270450> returned NULL without setting an error
[6060] Failed to execute script '__main__' due to unhandled exception!

```
(borg-1.1.16 is installed as distribution-pkg and works flawlessly)

**Comment by pete-dawgg** [comment]
Error reproducible "de_DE.iso885915@euro" with on Linux Mint Una for both vars with precompiled binary:
```
 LANG="de_DE.iso885915@euro" ./borg-linux64 
Traceback (most recent call last):
  File "borg/__main__.py", line 15, in <module>
  File "borg/archiver.py", line 5077, in main
  File "borg/archiver.py", line 4869, in get_args
  File "borg/archiver.py", line 4914, in parse_args
  File "borg/archiver.py", line 4339, in build_parser
  File "borg/helpers/parseformat.py", line 733, in keys_help
  File "borg/helpers/parseformat.py", line 727, in available_keys
  File "borg/helpers/parseformat.py", line 797, in get_item_data
SystemError: <cyfunction PropDict._make_property.<locals>._get at 0x7f6b2eb17450> returned NULL without setting an error
[4468] Failed to execute script '__main__' due to unhandled exception!
```

and
```
LC_ALL="de_DE.iso885915@euro" ./borg-linux64 
Traceback (most recent call last):
  File "borg/__main__.py", line 15, in <module>
  File "borg/archiver.py", line 5077, in main
  File "borg/archiver.py", line 4869, in get_args
  File "borg/archiver.py", line 4914, in parse_args
  File "borg/archiver.py", line 4339, in build_parser
  File "borg/helpers/parseformat.py", line 733, in keys_help
  File "borg/helpers/parseformat.py", line 727, in available_keys
  File "borg/helpers/parseformat.py", line 797, in get_item_data
SystemError: <cyfunction PropDict._make_property.<locals>._get at 0x7fcd6cc1d450> returned NULL without setting an error
[5027] Failed to execute script '__main__' due to unhandled exception!
```
However, it does NOT occur with de_DE.iso88591
BUT to (re)produce this error i had to manually generate the ISO-locales; w/out that system is utf8-only.

(borg-1.1.15 is installed as distribution-pkg and works flawlessly)

**Comment by ThomasWaldmann** [comment]
So, is this bug only about the fat binaries produced with pyinstaller (from github releases page) or does it also happen when using borg 1.2.0 from source or from a normal linux dist package?

**Comment by pete-dawgg** [comment]
i installed borgbackup 1.2.0-1 on debian bullseye as distro-package from testing ([bookworm:   https://packages.debian.org/bookworm/borgbackup   )](https://packages.debian.org/bookworm/borgbackup)  and the error is reproducible (and seems to have sth. to do with @euro):
LANG=de_DE: works
LANG=de_DE.iso88591: works
LANG=de_DE.iso885915@euro: ERROR:
```
Traceback (most recent call last):
  File "/usr/bin/borg", line 33, in <module>
    sys.exit(load_entry_point('borgbackup==1.2.0', 'console_scripts', 'borg')())
  File "/usr/lib/python3/dist-packages/borg/archiver.py", line 5077, in main
    args = archiver.get_args(sys.argv, os.environ.get('SSH_ORIGINAL_COMMAND'))
  File "/usr/lib/python3/dist-packages/borg/archiver.py", line 4869, in get_args
    result = self.parse_args(argv[1:])
  File "/usr/lib/python3/dist-packages/borg/archiver.py", line 4914, in parse_args
    parser = self.build_parser()
  File "/usr/lib/python3/dist-packages/borg/archiver.py", line 4339, in build_parser
    """) + ItemFormatter.keys_help()
  File "/usr/lib/python3/dist-packages/borg/helpers/parseformat.py", line 733, in keys_help
    keys = cls.available_keys()
  File "/usr/lib/python3/dist-packages/borg/helpers/parseformat.py", line 727, in available_keys
    keys.extend(formatter.get_item_data(fake_item).keys())
  File "/usr/lib/python3/dist-packages/borg/helpers/parseformat.py", line 797, in get_item_data
    source = item.get('source', '')
SystemError: <cyfunction PropDict._make_property.<locals>._get at 0x7f05655cf1e0> returned NULL without setting an error
```
LANG=de_DE@euro ERROR:
```
Traceback (most recent call last):
  File "/usr/bin/borg", line 33, in <module>
    sys.exit(load_entry_point('borgbackup==1.2.0', 'console_scripts', 'borg')())
  File "/usr/lib/python3/dist-packages/borg/archiver.py", line 5077, in main
    args = archiver.get_args(sys.argv, os.environ.get('SSH_ORIGINAL_COMMAND'))
  File "/usr/lib/python3/dist-packages/borg/archiver.py", line 4869, in get_args
    result = self.parse_args(argv[1:])
  File "/usr/lib/python3/dist-packages/borg/archiver.py", line 4914, in parse_args
    parser = self.build_parser()
  File "/usr/lib/python3/dist-packages/borg/archiver.py", line 4339, in build_parser
    """) + ItemFormatter.keys_help()
  File "/usr/lib/python3/dist-packages/borg/helpers/parseformat.py", line 733, in keys_help
    keys = cls.available_keys()
  File "/usr/lib/python3/dist-packages/borg/helpers/parseformat.py", line 727, in available_keys
    keys.extend(formatter.get_item_data(fake_item).keys())
  File "/usr/lib/python3/dist-packages/borg/helpers/parseformat.py", line 797, in get_item_data
    source = item.get('source', '')
SystemError: <cyfunction PropDict._make_property.<locals>._get at 0x7f5c3470d1e0> returned NULL without setting an error
```
Results for the error-causing locales above with fat precompiled binary.
```
Traceback (most recent call last):
  File "borg/__main__.py", line 15, in <module>
  File "borg/archiver.py", line 5077, in main
  File "borg/archiver.py", line 4869, in get_args
  File "borg/archiver.py", line 4914, in parse_args
  File "borg/archiver.py", line 4339, in build_parser
  File "borg/helpers/parseformat.py", line 733, in keys_help
  File "borg/helpers/parseformat.py", line 727, in available_keys
  File "borg/helpers/parseformat.py", line 797, in get_item_data
SystemError: <cyfunction PropDict._make_property.<locals>._get at 0x7f35c5617450> returned NULL without setting an error
[1616] Failed to execute script '__main__' due to unhandled exception!
```
No problems with utf8, but i did not check any en_-locales.

**Comment by ThomasWaldmann** [comment]
OK, so it's a general issue, not just pyinstaller binaries.

Python code (nothing special there):

https://github.com/borgbackup/borg/blob/1.2.0/src/borg/helpers/parseformat.py#L797

Cython code called by that:

https://github.com/borgbackup/borg/blob/1.2.0/src/borg/item.pyx#L107

I don't see how this leads to the error msg seen above. `cyfunction` sounds like a cython issue maybe?

**Comment by ThomasWaldmann** [comment]
@scoder can you have a look, please - any idea / hint?

**Comment by scoder** [comment]
What is `decode` here? Since this issue seems locale dependent, my *guess* is that `decode()` returns `NULL`, but Cython somehow doesn't know that it has to check the value as exception indicator.

**Comment by scoder** [comment]
You can also set the error level in `Cython.Compiler.Errors.LEVEL` to 0 (from 1). That might uncover a warning here.

**Comment by scoder** [comment]
I looked at the generated C code, it's a bug in Cython (both 0.29 and 3.0). Somehow, the `try ... except: raise ...` in the property's `_get()` function jumps to the wrong spot that resets the exception instead of propagating it.

https://github.com/cython/cython/issues/4675

**Comment by scoder** [comment]
So, back to step 1. It's not what I initially thought, and there's nothing obviously wrong with the C code that Cython generates here. Is there a way to reproduce this? @pete-dawgg, could you maybe run a debugger, or just fill up the `_get()` function with `print()` calls to see where it exits and what the variable values are?

**Comment by pete-dawgg** [comment]
@scoder "or just fill up the _get() function with print() calls to see where it exits and what the variable values are?"
Witrh clear instructions i guess i could. 
Which distro did you NOT get the error on? Does it have de_DE.iso885915@euro?

**Comment by scoder** [comment]
> What happens when you run `LC_ALL="de_DE@euro" borg` on your Ubuntu-system?

It prints the usage instructions, as normal.

**Comment by pete-dawgg** [comment]
Did a quick test with Debian Bookworm in a chroot. Before being able to use and compile borg in a venv a lot of (newer) python-pkgs had to be installed as system-pkgs. This made the error go away except with LC_ALL="de_DE.iso885915@euro"

```
(borg-env) gen2 /tmp/borg-env/borgbackup-1.2.0 # LC_ALL="de_DE.iso885915@euro" borg
KEY: mode
VALUE: 0
KEY: source
KeyError source attribute source not found
Traceback (most recent call last):
  File "/tmp/borg-env/bin/borg", line 8, in <module>
    sys.exit(main())
  File "/tmp/borg-env/lib/python3.9/site-packages/borg/archiver.py", line 5077, in main
    args = archiver.get_args(sys.argv, os.environ.get('SSH_ORIGINAL_COMMAND'))
  File "/tmp/borg-env/lib/python3.9/site-packages/borg/archiver.py", line 4869, in get_args
    result = self.parse_args(argv[1:])
  File "/tmp/borg-env/lib/python3.9/site-packages/borg/archiver.py", line 4914, in parse_args
    parser = self.build_parser()
  File "/tmp/borg-env/lib/python3.9/site-packages/borg/archiver.py", line 4339, in build_parser
    """) + ItemFormatter.keys_help()
  File "/tmp/borg-env/lib/python3.9/site-packages/borg/helpers/parseformat.py", line 733, in keys_help
    keys = cls.available_keys()
  File "/tmp/borg-env/lib/python3.9/site-packages/borg/helpers/parseformat.py", line 727, in available_keys
    keys.extend(formatter.get_item_data(fake_item).keys())
  File "/tmp/borg-env/lib/python3.9/site-packages/borg/helpers/parseformat.py", line 797, in get_item_data
    source = item.get('source', '')
SystemError: <cyfunction PropDict._make_property.<locals>._get at 0x7f6130f14790> returned NULL without setting an error
```

LANG="de_DE.iso885915@euro" 
```
(borg-env) gen2 /tmp/borg-env/borgbackup-1.2.0 # LANG="de_DE.iso885915@euro" borg
KEY: mode
VALUE: 0
KEY: source
KeyError source attribute source not found
KEY: user
VALUE: b''
decode: <function safe_decode at 0x7f0d38ba14c0>
decoded value: 
KEY: uid
VALUE: 0
KEY: group
VALUE: b''
decode: <function safe_decode at 0x7f0d38ba14c0>
decoded value: 
KEY: gid
VALUE: 0
KEY: uid
VALUE: 0
KEY: gid
VALUE: 0
KEY: path
VALUE: b''
decode: <function safe_decode at 0x7f0d38ba14c0>
decoded value: 
KEY: path
VALUE: b''
decode: <function safe_decode at 0x7f0d38ba14c0>
decoded value: 
KEY: bsdflags
KeyError bsdflags attribute bsdflags not found
```

if it helps i can check in a "real" (non-chrooted) system later.

**Comment by pete-dawgg** [comment]
1st change with `raise err from None`:
```
(borg-env) 05:03:51 ice@gen2 /tmp/borg-env/borg-debug-src $ LC_ALL="de_DE.iso885915@euro" borg
KEY: mode
VALUE: 0
KEY: source
KeyError source attribute source not found
raising attribute source not found
Traceback (most recent call last):
  File "/tmp/borg-env/bin/borg", line 8, in <module>
    sys.exit(main())
  File "/tmp/borg-env/lib/python3.9/site-packages/borg/archiver.py", line 5077, in main
    args = archiver.get_args(sys.argv, os.environ.get('SSH_ORIGINAL_COMMAND'))
  File "/tmp/borg-env/lib/python3.9/site-packages/borg/archiver.py", line 4869, in get_args
    result = self.parse_args(argv[1:])
  File "/tmp/borg-env/lib/python3.9/site-packages/borg/archiver.py", line 4914, in parse_args
    parser = self.build_parser()
  File "/tmp/borg-env/lib/python3.9/site-packages/borg/archiver.py", line 4339, in build_parser
    """) + ItemFormatter.keys_help()
  File "/tmp/borg-env/lib/python3.9/site-packages/borg/helpers/parseformat.py", line 733, in keys_help
    keys = cls.available_keys()
  File "/tmp/borg-env/lib/python3.9/site-packages/borg/helpers/parseformat.py", line 727, in available_keys
    keys.extend(formatter.get_item_data(fake_item).keys())
  File "/tmp/borg-env/lib/python3.9/site-packages/borg/helpers/parseformat.py", line 797, in get_item_data
    source = item.get('source', '')
SystemError: <cyfunction PropDict._make_property.<locals>._get at 0x7fd0e417a790> returned NULL without setting an error
```

2nd change with `raise err` only:

```
(borg-env) 05:13:18 ice@gen2 /tmp/borg-env/borg-debug-src $ LC_ALL="de_DE.iso885915@euro" borg
KEY: mode
VALUE: 0
KEY: source
KeyError source attribute source not found
raising attribute source not found
Traceback (most recent call last):
  File "/tmp/borg-env/bin/borg", line 8, in <module>
    sys.exit(main())
  File "/tmp/borg-env/lib/python3.9/site-packages/borg/archiver.py", line 5077, in main
    args = archiver.get_args(sys.argv, os.environ.get('SSH_ORIGINAL_COMMAND'))
  File "/tmp/borg-env/lib/python3.9/site-packages/borg/archiver.py", line 4869, in get_args
    result = self.parse_args(argv[1:])
  File "/tmp/borg-env/lib/python3.9/site-packages/borg/archiver.py", line 4914, in parse_args
    parser = self.build_parser()
  File "/tmp/borg-env/lib/python3.9/site-packages/borg/archiver.py", line 4339, in build_parser
    """) + ItemFormatter.keys_help()
  File "/tmp/borg-env/lib/python3.9/site-packages/borg/helpers/parseformat.py", line 733, in keys_help
    keys = cls.available_keys()
  File "/tmp/borg-env/lib/python3.9/site-packages/borg/helpers/parseformat.py", line 727, in available_keys
    keys.extend(formatter.get_item_data(fake_item).keys())
  File "/tmp/borg-env/lib/python3.9/site-packages/borg/helpers/parseformat.py", line 797, in get_item_data
    source = item.get('source', '')
SystemError: <cyfunction PropDict._make_property.<locals>._get at 0x7f92097b1790> returned NULL without setting an error
```

**Comment by scoder** [comment]
Thanks for testing. Could you validate that the C code that I posted is also what you see in the file `src/borg/item.c`, around line 4500? And that it says "generated by Cython 0.29.28" at the top of the file? (If there is no such file, it's probably because "pip" did its build in a different directory. In that case, see if it built a "borg-1.2.0.tar.gz" file under "dist/" and check if the file is included in it.)

From what I can see, this can't be a Cython bug. It looks like you did all your tests with Python 3.9. Could you also try Python 3.8 or 3.10, to see if they show the same failure?

**Comment by pete-dawgg** [comment]
it does say:` /* Generated by Cython 0.29.28 */` at the top of the file (using the "original" item.pyx)
then:
 ```
4489   /* "borg/item.pyx":107
 4490  *         attr_error_msg = "attribute %s not found" % key
 4491  * 
 4492  *         def _get(self):             # <<<<<<<<<<<<<<
 4493  *             try:
 4494  *                 value = self._dict[key]
 4495  */
 4496 
 4497   /* function exit code */
 4498   __pyx_L1_error:;
 4499   __Pyx_XDECREF(__pyx_t_4);
 4500   __Pyx_XDECREF(__pyx_t_5);
 4501   __Pyx_XDECREF(__pyx_t_7);
 4502   __Pyx_XDECREF(__pyx_t_8);
 4503   __Pyx_AddTraceback("borg.item.PropDict._make_property._get", __pyx_clineno, __pyx_lineno, __pyx_filename);
 4504   __pyx_r = NULL;
 4505   __pyx_L0:;
 4506   __Pyx_XDECREF(__pyx_v_value);
 4507   __Pyx_XGIVEREF(__pyx_r);
 4508   __Pyx_RefNannyFinishContext();
 4509   return __pyx_r;
 4510 }
```

**Comment by pete-dawgg** [comment]
Python 3.8.12:
```
(borg-env) 04:41:14 ice@gen2 /tmp/borg-env/borg $ LC_ALL="de_DE.iso885915@euro" borg -v
Traceback (most recent call last):
  File "/tmp/borg-env/bin/borg", line 8, in <module>
    sys.exit(main())
  File "/tmp/borg-env/lib/python3.8/site-packages/borg/archiver.py", line 5077, in main
    args = archiver.get_args(sys.argv, os.environ.get('SSH_ORIGINAL_COMMAND'))
  File "/tmp/borg-env/lib/python3.8/site-packages/borg/archiver.py", line 4869, in get_args
    result = self.parse_args(argv[1:])
  File "/tmp/borg-env/lib/python3.8/site-packages/borg/archiver.py", line 4914, in parse_args
    parser = self.build_parser()
  File "/tmp/borg-env/lib/python3.8/site-packages/borg/archiver.py", line 4339, in build_parser
    """) + ItemFormatter.keys_help()
  File "/tmp/borg-env/lib/python3.8/site-packages/borg/helpers/parseformat.py", line 733, in keys_help
    keys = cls.available_keys()
  File "/tmp/borg-env/lib/python3.8/site-packages/borg/helpers/parseformat.py", line 727, in available_keys
    keys.extend(formatter.get_item_data(fake_item).keys())
  File "/tmp/borg-env/lib/python3.8/site-packages/borg/helpers/parseformat.py", line 797, in get_item_data
    source = item.get('source', '')
SystemError: <cyfunction PropDict._make_property.<locals>._get at 0x7f96de935ee0> returned NULL without setting an error
```
Python 3.10.2:

```
/tmp/borg-env $ LC_ALL="de_DE.iso885915@euro" borg -v
Traceback (most recent call last):
  File "/tmp/borg-env/bin/borg", line 8, in <module>
    sys.exit(main())
  File "/tmp/borg-env/lib/python3.10/site-packages/borg/archiver.py", line 5077, in main
    args = archiver.get_args(sys.argv, os.environ.get('SSH_ORIGINAL_COMMAND'))
  File "/tmp/borg-env/lib/python3.10/site-packages/borg/archiver.py", line 4869, in get_args
    result = self.parse_args(argv[1:])
  File "/tmp/borg-env/lib/python3.10/site-packages/borg/archiver.py", line 4914, in parse_args
    parser = self.build_parser()
  File "/tmp/borg-env/lib/python3.10/site-packages/borg/archiver.py", line 4339, in build_parser
    """) + ItemFormatter.keys_help()
  File "/tmp/borg-env/lib/python3.10/site-packages/borg/helpers/parseformat.py", line 733, in keys_help
    keys = cls.available_keys()
  File "/tmp/borg-env/lib/python3.10/site-packages/borg/helpers/parseformat.py", line 727, in available_keys
    keys.extend(formatter.get_item_data(fake_item).keys())
  File "/tmp/borg-env/lib/python3.10/site-packages/borg/helpers/parseformat.py", line 797, in get_item_data
    source = item.get('source', '')
SystemError: <cyfunction PropDict._make_property.<locals>._get at 0x7f24c47c8ba0> returned NULL without setting an exception
```

**Comment by pete-dawgg** [comment]
wasn't sure how to add the cflags to pip install so i added
```
import sysconfig
extra_compile_args = sysconfig.get_config_var('CFLAGS').split()
extra_compile_args += ["-O0", "-ggdb"]
```
to setup.py  line 28.
found the code you posted, in my case it's line 4400.

the command you posted crashed after run:

```
Reading symbols from python...
(No debugging symbols found in python)
(gdb) break src/item.c:4400
No symbol table is loaded.  Use the "file" command.
Make breakpoint pending on future shared library load? (y or [n]) 
(gdb) run
Starting program: /tmp/borg-env/bin/python -m borg -V
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib64/libthread_db.so.1".
Traceback (most recent call last):
  File "/usr/lib/python3.10/runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/lib/python3.10/runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "/tmp/borg-env/lib/python3.10/site-packages/borg/__main__.py", line 15, in <module>
    main()
  File "/tmp/borg-env/lib/python3.10/site-packages/borg/archiver.py", line 5077, in main
    args = archiver.get_args(sys.argv, os.environ.get('SSH_ORIGINAL_COMMAND'))
  File "/tmp/borg-env/lib/python3.10/site-packages/borg/archiver.py", line 4869, in get_args
    result = self.parse_args(argv[1:])
  File "/tmp/borg-env/lib/python3.10/site-packages/borg/archiver.py", line 4914, in parse_args
    parser = self.build_parser()
  File "/tmp/borg-env/lib/python3.10/site-packages/borg/archiver.py", line 4339, in build_parser
    """) + ItemFormatter.keys_help()
  File "/tmp/borg-env/lib/python3.10/site-packages/borg/helpers/parseformat.py", line 733, in keys_help
    keys = cls.available_keys()
  File "/tmp/borg-env/lib/python3.10/site-packages/borg/helpers/parseformat.py", line 727, in available_keys
    keys.extend(formatter.get_item_data(fake_item).keys())
  File "/tmp/borg-env/lib/python3.10/site-packages/borg/helpers/parseformat.py", line 797, in get_item_data
    source = item.get('source', '')
SystemError: <cyfunction PropDict._make_property.<locals>._get at 0x7ffff5618ee0> returned NULL without setting an exception
[Inferior 1 (process 17782) exited with code 01]
(gdb) n
The program is not being run.
(gdb) 
```

**Comment by ThomasWaldmann** [comment]
any progress here? will the upcoming 0.29.29 cython release fix it?

**Comment by scoder** [comment]
No, I still don't have enough information to know what is going on here.

@pete-dawgg, the breakpoint wasn't hit. I think you need to write `break src/borg/item.c:[LINE]` instead of ` break src/item.c:[LINE]`, as you showed. It should stop at that line before it crashes.

**Comment by scoder** [comment]
> any progress here? will the upcoming 0.29.29 cython release fix it?

I doesn't seem like anyone besides the OP has managed to reproduce this behaviour yet. That makes it difficult to guess in which part of the software stack (Hardware … OS … Python … Cython … Borg) the problem may arise. At this point, it's at least unlikely that it's a problem caused by Cython.

**Comment by ThomasWaldmann** [comment]
I can reproduce the issue on Ubuntu 20.04 amd64, steps I needed:

```
dpkg-reconfigure -plow locales  # make sure all the de_DE* locales are selected
LC_ALL="de_DE.iso-8859-15@euro" ./borg11  # 1.1.17 binary, works ok
LC_ALL="de_DE.iso-8859-15@euro" ./borg12 # 1.2.0 linux64 binary, crashes
LC_ALL="de_DE.iso-8859-15@euro" ./borg12o # 1.2.0 oldlinux64 binary, crashes
LC_ALL="de_DE@euro" ... # same as above
LC_ALL="de_DE" ... # everything works
LC_ALL="de_DE.UTF-8" ... # everything works
```

**Comment by ThomasWaldmann** [comment]
Also did a git checkout of 1.2.0 and installed it into virtual env:
- with cython 0.29.25, 0.29.26, 0.29.27, 0.29.28: crashes
- with cython 0.29.23, 0.29.24 and 0.29.29, 0.29.30: works

I switched back and forth multiple times. Changing the cython version and doing a clean build is enough to make the issue disappear and come back.

borg 1.1.17 just worked because we did not use the defect cython versions when releasing it.
Building it with the problematic versions makes it crash also, building with the latest version makes it work also.

So, summarizing: cython 0.29.25 .. 0.29.28 have this issue. Before and after these, it works.

@scoder FYI: looks like somebody fixed it already in cython.

Closing this: not a borg issue and we'll use the latest cython 0.29.x for next releases anyway.
